inspirehep/modules/workflows/workflows/article.py

```diff
@@ -317,26 +262,180 @@
+                message="Submission halted for curator approval.",
+            )
         ],
-    ),
+        # record not relevant
+        [
+            reject_record("Article automatically rejected"),
+            mark('approved', False),
+            save_workflow,
+            stop_processing,
+        ]
+    )
+]
+
+
+STORE_RECORD = [
+    store_record
+]
+
+
+MARK_IF_MATCH_IN_HOLDINGPEN = [
     IF(
-        pending_in_holding_pen,
+        match_non_completed_wf_in_holdingpen,
         [
             mark('already-in-holding-pen', True),
-        ],
+            save_workflow,
+        ]
+    ),
+    IF(
+        match_previously_rejected_wf_in_holdingpen,
+        [
+            mark('previously_rejected', True),
+            save_workflow,
+        ]
+    )
+]
+
+
+ERROR_WITH_UNEXPECTED_WORKFLOW_PATH = [
+    mark('unexpected-workflow-path', True),
+    error_workflow('Unexpected workflow path.'),
+    save_workflow,
+]
+
+
+# Currently we handle harvests as if all were arxiv, that will have to change.
+PROCESS_HOLDINGPEN_MATCH_ARXIV = [
+    IF(
+        is_marked('previously_rejected'),
+        IF(
+            holdingpen_match_with_same_source,
+            [
+                mark('approved', False),  # auto-reject
+                mark('previously_rejected', True),
+                save_workflow,
+                stop_processing,
+            ]
+        )
     ),
+
+    IF(
+        is_marked('already-in-holding-pen'),
+        IF(
+            holdingpen_match_with_same_source,
+            # stop the matched wf and continue this one
+            [
+                stop_matched_holdingpen_wfs,
+                mark('stopped-matched-holdingpen-wf', True),
+                save_workflow
+            ],
+        )
+        # else, it's an update from another source
+    )
+]
+
+
+PROCESS_HOLDINGPEN_MATCH_SUBMISSION = [
+    IF(
+        is_marked('already-in-holding-pen'),
+        IF_ELSE(
+            holdingpen_match_with_same_source,
+            # form should detect this double submission
+            ERROR_WITH_UNEXPECTED_WORKFLOW_PATH,
+
+            # stop the matched wf and continue this one
+            [
+                stop_matched_holdingpen_wfs,
+                mark('stopped-matched-holdingpen-wf', True),
+                save_workflow
+            ],
+
+        )
+    )
+]
+
+
+PROCESS_HOLDINGPEN_MATCHES = [
     IF_ELSE(
+        is_submission,
+        PROCESS_HOLDINGPEN_MATCH_SUBMISSION,
+
+        # not a submission, harvest
+        IF_ELSE(
+            is_arxiv_paper,
+            PROCESS_HOLDINGPEN_MATCH_ARXIV,
+
+            # We don't handle yet publisher harvests
+            ERROR_WITH_UNEXPECTED_WORKFLOW_PATH,
+        ),
+
+    )
+]
+
+
+MARK_IF_UPDATE = [
+    IF(
+        article_exists,
+        [
+            mark('is-update', True),
+            save_workflow,
+        ]
+    )
+]
+
+
+STOP_IF_ALREADY_HARVESTED_ON_LEGACY = [
+    IF(
+        already_harvested,
```

Comment:
already_harvested should check too the day limit until merger is in (or in another better named step here too).